File: UltraMesh 2d animation/Core/Tools/RotateTool.swift
import Foundation
import simd

final class RotateTool: Tool {
    let type: ActiveTool = .rotate

    private var activeID: UUID?
    private var activeBoneID: UUID?
    /// Every selected bone, parents first — see `selectedBonesInChainOrder`.
    private var boneDragOrder: [UUID] = []
    /// Where each of them started, so the drag is relative. One start value for
    /// the lot would snap every bone onto the active one's angle.
    private var boneStartRotations: [UUID: Float] = [:]
    private var startRotation: Float = 0
    private var startAngle: Float = 0
    private var boneStart: SIMD2<Float> = .zero
    private var boneLength: Float = 0
    private var settleTarget: Float?
    private var settleID: UUID?
    private var mouseDownPosition: SIMD2<Float>?
    private var didDrag = false

    private let settleFactor: Float = 0.22
    private let snapStepDegrees: Float = 15
    private let dragThreshold: Float = 4

    private func ensureDragStarted(currentPosition: SIMD2<Float>) -> Bool {
        if didDrag { return true }
        guard let down = mouseDownPosition else { return false }
        if simd_distance(currentPosition, down) >= dragThreshold {
            didDrag = true
            return true
        }
        return false
    }

    func onMouseDown(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        guard input.activeHandle == .rotateRing else { return }
        scene.beginInteraction()
        mouseDownPosition = input.position
        didDrag = false
        if let boneID = scene.selectedBoneID,
           let segment = scene.skeleton.lineSegment(for: boneID) {
            activeBoneID = boneID
            boneStart = segment.start
            boneLength = max(simd_distance(segment.start, segment.end), 12)
            let vector = input.startPosition - segment.start
            startAngle = atan2(vector.y, vector.x)
            startRotation = scene.skeleton.worldRotation(for: boneID) ?? 0
            // The whole selection turns, not only the bone the ring is drawn
            // on. Depth order because `setBoneRotation` converts a world angle
            // against the parent's CURRENT world rotation: a child written
            // before its parent is measured against the old angle and then
            // inherits the parent's turn on top, so it turns twice.
            boneDragOrder = scene.selectedBonesInDepthOrder
            boneStartRotations = Dictionary(uniqueKeysWithValues: boneDragOrder.compactMap { id in
                scene.skeleton.worldRotation(for: id).map { (id, $0) }
            })
            return
        }
        let hit = scene.selectedImageID ?? ToolUtilities.hitTestScreen(
            screenPoint: input.screenPosition,
            viewSize: input.viewSize,
            scene: scene,
            assets: assets,
            camera: input.camera
        )
        guard let id = hit, let image = scene.image(for: id) else { return }
        scene.selectedImageID = id
        activeID = id
        startRotation = image.rotation
        if simd_length(image.rotation3D) > 0.0001 {
            scene.updateImage(id: id) { image in
                image.rotation3D = .zero
            }
        }
        settleTarget = nil
        settleID = nil
        let vector = input.startPosition - image.position
        startAngle = atan2(vector.y, vector.x)
    }

    func onMouseDrag(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        guard ensureDragStarted(currentPosition: input.position) else { return }
        if let boneID = activeBoneID {
            let vector = input.position - boneStart
            let angle = atan2(vector.y, vector.x)
            let delta = angle - startAngle
            var rotation = startRotation + delta
            if input.isShiftPressed {
                rotation = ToolUtilities.snapAngle(rotation, stepDegrees: snapStepDegrees)
            }
            // Taken from the ACTIVE bone after snapping, so Shift snaps that
            // bone to the grid and the rest of the selection follows rigidly.
            // Snapping each bone on its own would pull the group apart at every
            // step boundary.
            let appliedDelta = rotation - startRotation
            guard !boneDragOrder.isEmpty else {
                scene.setBoneRotation(id: boneID, worldAngle: rotation)
                return
            }
            for id in boneDragOrder {
                guard let start = boneStartRotations[id] else { continue }
                scene.setBoneRotation(id: id, worldAngle: start + appliedDelta)
            }
            return
        }
        guard let id = activeID, let image = scene.image(for: id) else { return }
        let vector = input.position - image.position
        let angle = atan2(vector.y, vector.x)
        let delta = angle - startAngle
        var rotation = startRotation + delta
        if input.isShiftPressed {
            rotation = ToolUtilities.snapAngle(rotation, stepDegrees: snapStepDegrees)
        }
        scene.setImageRotation(id: id, rotation: rotation)
    }

    func onMouseUp(input: ToolInput, scene: SceneManager, assets: AssetManager) {
        scene.endInteraction()
        defer {
            activeBoneID = nil
            boneDragOrder = []
            boneStartRotations = [:]
            activeID = nil
            mouseDownPosition = nil
            didDrag = false
        }
        guard didDrag else { return }
        if let boneID = activeBoneID {
            // A key for each bone that moved. Committing only the active one
            // recorded a pose in which the rest of the selection had turned on
            // screen and nowhere in the animation.
            for id in (boneDragOrder.isEmpty ? [boneID] : boneDragOrder) {
                scene.commitKeyframe(for: id, property: .rotate)
            }
            return
        }
        guard let id = activeID, let image = scene.image(for: id) else { return }
        let raw = image.rotation
        let target = input.isShiftPressed ? ToolUtilities.snapAngle(raw, stepDegrees: snapStepDegrees) : raw
        settleTarget = target
        settleID = id
        scene.commitKeyframe(for: id, property: .rotate)
    }

    func update(scene: SceneManager, assets: AssetManager) {
        guard let id = settleID, let target = settleTarget, let image = scene.image(for: id) else { return }
        let current = image.rotation
        let next = current + (target - current) * settleFactor
        scene.setImageRotation(id: id, rotation: next)
        if abs(next - target) < 0.001 {
            scene.setImageRotation(id: id, rotation: target)
            settleTarget = nil
            settleID = nil
        }
    }
}

File: UMeshCore/src/Editor/Tools/RotateTool.cpp
#include "umeshcore/Editor/Tools/RotateTool.h"

#include <cmath>
#include <variant>

namespace umeshcore {

bool RotateTool::ensureDragStarted(const Vec2& currentPosition) {
    if (didDrag_) return true;
    if (!mouseDownPosition_.has_value()) return false;
    if (length(currentPosition - *mouseDownPosition_) >= kDragThreshold) {
        didDrag_ = true;
        return true;
    }
    return false;
}

void RotateTool::onMouseDown(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& imageHitTest, float /*hitScale*/,
    bool /*touchOptimized*/) {
    if (!input.activeHandle.has_value() || !std::holds_alternative<RotateRingHandle>(*input.activeHandle)) return;
    scene.beginInteraction();
    mouseDownPosition_ = input.position;
    didDrag_ = false;

    if (scene.selectedBoneID.has_value()) {
        const auto segment = scene.skeleton.lineSegment(*scene.selectedBoneID);
        if (segment.has_value()) {
            activeBoneID_ = scene.selectedBoneID;
            boneStart_ = segment->start;
            boneLength_ = std::max(length(segment->end - segment->start), 12.0f);
            const Vec2 vector = input.startPosition - segment->start;
            startAngle_ = std::atan2(vector.y, vector.x);
            const auto worldRotation = scene.skeleton.worldRotation(*scene.selectedBoneID);
            startRotation_ = worldRotation.has_value() ? *worldRotation : 0.0f;
            // The whole selection turns, not only the bone the ring is drawn
            // on. Depth order because setBoneRotation converts a world angle
            // against the parent's CURRENT world rotation.
            boneDragOrder_ = scene.selectedBonesInDepthOrder();
            boneStartRotations_.clear();
            for (Uuid id : boneDragOrder_) {
                if (const auto r = scene.skeleton.worldRotation(id)) boneStartRotations_[id] = *r;
            }
            return;
        }
    }

    const std::optional<ImageHit> imageHit =
        imageHitTest ? imageHitTest(input.screenPosition, input.viewSize, input.camera) : std::nullopt;
    const std::optional<Uuid> hit =
        scene.selectedImageID.has_value() ? scene.selectedImageID
        : (imageHit.has_value() ? std::optional<Uuid>(imageHit->id) : std::nullopt);
    if (!hit.has_value()) return;
    SceneImage* image = scene.image(*hit);
    if (image == nullptr) return;

    scene.selectedImageID = *hit;
    activeID_ = *hit;
    startRotation_ = image->rotation;
    if (length(image->rotation3D) > 0.0001f) {
        scene.setImageRotation3D(*hit, Vec3::zero());
        image = scene.image(*hit); // setImageRotation3D may have moved the vector; re-fetch.
    }
    settleTarget_ = std::nullopt;
    settleID_ = std::nullopt;
    const Vec2 vector = input.startPosition - image->position;
    startAngle_ = std::atan2(vector.y, vector.x);
}

void RotateTool::onMouseDrag(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool /*touchOptimized*/) {
    if (!ensureDragStarted(input.position)) return;

    if (activeBoneID_.has_value()) {
        const Vec2 vector = input.position - boneStart_;
        const float angle = std::atan2(vector.y, vector.x);
        const float delta = angle - startAngle_;
        float rotation = startRotation_ + delta;
        if (input.isShiftPressed) rotation = ToolUtilities::snapAngle(rotation, kSnapStepDegrees);
        // Taken from the ACTIVE bone after snapping, so Shift snaps that
        // bone to the grid and the rest of the selection follows rigidly.
        const float appliedDelta = rotation - startRotation_;
        if (boneDragOrder_.empty()) {
            scene.setBoneRotation(*activeBoneID_, rotation);
            return;
        }
        for (Uuid id : boneDragOrder_) {
            auto it = boneStartRotations_.find(id);
            if (it == boneStartRotations_.end()) continue;
            scene.setBoneRotation(id, it->second + appliedDelta);
        }
        return;
    }

    if (!activeID_.has_value()) return;
    SceneImage* image = scene.image(*activeID_);
    if (image == nullptr) return;
    const Vec2 vector = input.position - image->position;
    const float angle = std::atan2(vector.y, vector.x);
    const float delta = angle - startAngle_;
    float rotation = startRotation_ + delta;
    if (input.isShiftPressed) rotation = ToolUtilities::snapAngle(rotation, kSnapStepDegrees);
    scene.setImageRotation(*activeID_, rotation);
}

void RotateTool::onMouseUp(
    const ToolInput& input, EditorScene& scene, const ImageHitTestFn& /*imageHitTest*/, float /*hitScale*/,
    bool /*touchOptimized*/) {
    scene.endInteraction();

    const std::optional<Uuid> boneID = activeBoneID_;
    const std::optional<Uuid> imageID = activeID_;
    const std::vector<Uuid> boneDragOrder = boneDragOrder_;
    const bool didDrag = didDrag_;

    activeBoneID_ = std::nullopt;
    boneDragOrder_.clear();
    boneStartRotations_.clear();
    activeID_ = std::nullopt;
    mouseDownPosition_ = std::nullopt;
    didDrag_ = false;

    if (!didDrag) return;

    if (boneID.has_value()) {
        // A key for each bone that moved -- committing only the active one
        // would record a pose where the rest of the selection turned on
        // screen and nowhere in the animation.
        const std::vector<Uuid> ids = boneDragOrder.empty() ? std::vector<Uuid>{*boneID} : boneDragOrder;
        for (Uuid id : ids) {
            scene.commitKeyframe(id, AnimationTrackProperty::Rotate);
        }
        return;
    }

    if (!imageID.has_value()) return;
    SceneImage* image = scene.image(*imageID);
    if (image == nullptr) return;
    const float raw = image->rotation;
    const float target = input.isShiftPressed ? ToolUtilities::snapAngle(raw, kSnapStepDegrees) : raw;
    settleTarget_ = target;
    settleID_ = imageID;
    scene.commitKeyframe(*imageID, AnimationTrackProperty::Rotate);
}

void RotateTool::update(EditorScene& scene) {
    if (!settleID_.has_value() || !settleTarget_.has_value()) return;
    SceneImage* image = scene.image(*settleID_);
    if (image == nullptr) return;
    const float current = image->rotation;
    const float next = current + (*settleTarget_ - current) * kSettleFactor;
    scene.setImageRotation(*settleID_, next);
    if (std::abs(next - *settleTarget_) < 0.001f) {
        scene.setImageRotation(*settleID_, *settleTarget_);
        settleTarget_ = std::nullopt;
        settleID_ = std::nullopt;
    }
}

} // namespace umeshcore
